CRM/Member/BAO/Membership.php

```diff
@@ -2377,42 +2377,34 @@ private static function processOverriddenUntilDateMembership($membership) {
   }
 
   /**
-   * Returns the membership types for a particular contact
-   * who has lifetime membership without end date.
+   * Returns the membership types for a contact, optionally filtering to lifetime memberships only.
    *
    * @param int $contactID
    * @param bool $isTest
    * @param bool $onlyLifeTime
    *
    * @return array
    */
-  public static function getAllContactMembership($contactID, $isTest = FALSE, $onlyLifeTime = FALSE) {
+  public static function getAllContactMembership($contactID, $isTest = FALSE, $onlyLifeTime = FALSE) : array {
     $contactMembershipType = [];
     if (!$contactID) {
       return $contactMembershipType;
     }
 
-    $dao = new CRM_Member_DAO_Membership();
-    $dao->contact_id = $contactID;
-    $pendingStatusId = array_search('Pending', CRM_Member_PseudoConstant::membershipStatus());
-    $dao->whereAdd("status_id != $pendingStatusId");
-
-    if ($isTest) {
-      $dao->is_test = $isTest;
-    }
-    else {
-      $dao->whereAdd('is_test IS NULL OR is_test = 0');
-    }
+    $membershipQuery = \Civi\Api4\Membership::get(FALSE)
+      ->addWhere('contact_id', '=', $contactID)
+      ->addWhere('status_id:name', '<>', 'Pending')
+      ->addWhere('is_test', '=', $isTest)
```

Comment:
@MegaphoneJon should this not be done on conditional basis like before?  or is it the point that is_test cannot be NULL anymore and FALSE will be converted to 0 by APIv4?

Reply:
Yes - there's a default set in the DAO which didn't use to exist.

That's actually true for a lot of fields (`civicrm_contact.is_deceased` being a [particular pain point](https://civicrm.org/extensions/zombie-check) once upon a time.  Removing these conditionals as we find them is a good idea IMO.

Reply:
@shaneonabike Just to manage expectations - lots of places in their code implement lifetime membership checks, and only one of those places is getting fixed by this PR (the check that comes on public contribution pages).

Once this is merged we can replace the lifetime membership checks in other parts of the code with this one, once we agree this is correct and reliable.  So this might only be the first step toward fixing your issue.